File: frontend/bagshop/src/service/ProductService.js
import axiosClient from "./AxiosConfig";

const ProductService = {
  getAll() {
    return axiosClient.get("product");
  },
  getProductsWithPaging(pageNumber, pageSize, sortBy, sortDir) {
    return axiosClient.get(
      `product/getAllProductsWithPaging/${pageNumber}/${pageSize}/${sortBy}/${sortDir}`,
    );
  },
  getProductById(productId) {
    return axiosClient.get(`product/${productId}`);
  },
  getProductByCategoryId(categoryId) {
    return axiosClient.get(`product/category/${categoryId}`);
  },
  createProduct(productRequest) {
    return axiosClient.post("product/create", productRequest);
  },

  createVariant(productId, productVariantRequest) {
    return axiosClient.post(
      `product/${productId}/variants`,
      productVariantRequest,
    );
  },
  updateProduct(productId, productRequest) {
    return axiosClient.put(`product/update/${productId}`, productRequest);
  },

  updateVariant(productId, variantId, productVariantRequest) {
    return axiosClient.put(
      `product/${productId}/variants/${variantId}`,
      productVariantRequest,
    );
  },
  deleteProduct(productId) {
    return axiosClient.delete(`product/${productId}`);
  },
  deleteVariant(variantId) {
    return axiosClient.delete(`product/variants/${variantId}`);
  },
};

export default ProductService;

File: backend/src/main/java/bag/service/product/ProductService.java
package bag.service.product;

import bag.modal.dto.ProductDto;
import bag.modal.dto.ProductVariantDto;
import bag.modal.request.ProductRequest;
import bag.modal.request.ProductVariantRequest;
import org.springframework.data.domain.Page;

import java.util.List;

public interface ProductService {
    Page<ProductDto> getAllProductsWithPaging(int page, int size, String sortBy, String sortDir);

    List<ProductDto> getAllProductsWithoutPaging();
    List<ProductDto> getByCategoryId(int categoryId);
    List<ProductDto> getProductByRangePrice(double minPrice, double maxPrice );
    ProductDto getProductById(int productId);
    List<ProductDto> getProductsHighToLow();

    ProductDto addProduct(ProductRequest request);

    ProductDto updateProduct(ProductRequest request, int id);
    ProductDto deleteProductById(int productId);

    //variant
    ProductVariantDto createVariant(int productId, ProductVariantRequest request);

    ProductVariantDto updateVariant( ProductVariantRequest request, int id) ;

    ProductVariantDto getVariantBySku(String sku);

    ProductVariantDto getVariantByColorAndSize(int productId, String color, String size);


    ProductVariantDto deleteVariant(int variantId);
}
